cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -121,28 +133,71 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
+   * directory hash to ensure the contents are unchanged.
+   *
+   * @param relativePath The path relative to the repository base path (returned by
+   *                     {@link RepositoryManager#getRepositoryRoot()}) on the filesystem.
+   * @param commitID     The commit ID hash for which to get the file hash.
+   * @return The git file hash of the requested file.
+   * @throws IOException              when file or commit isn't found, there are circular symlinks or other file IO
+   *                                  errors.
+   * @throws IllegalArgumentException when the provided file isn't a regular file, eg: directory.
+   * @throws NotFoundException        when the file isn't committed to Git.
+   * @throws IllegalStateException    when {@link RepositoryManager#cloneRemote()} isn't called before this.
+   */
+  public String getFileHash(Path relativePath, String commitID) throws IOException, NotFoundException {
+    validateInitialized();
+    // Git considers symlinks as separate files. So the hash of the symlink will be different from the target file.
+    // Resolve path in case of symlinks. This ensures the hash of the target file is returned.
+    Path realPath = getRepositoryRoot().resolve(relativePath).toRealPath();
+    // We should not be comparing hashes of directories. If the directory contains symlinks, we can't depend on the
+    // directory hash to ensure the contents are unchanged.
+    if (!Files.isRegularFile(realPath)) {
+      throw new IllegalArgumentException(String.format("Path %s doesn't refer to a regular file.",
+                                                       realPath.toAbsolutePath()));
+    }
+    Path realRelativePath = getRepositoryRoot().relativize(realPath);
+    ObjectId lastCommitId = ObjectId.fromString(commitID);
+    // Each commit points to a tree that contains the files/directories as nodes. The ObjectId (hashes) of the nodes
+    // represent the state of a file at that commit.
+    RevCommit commit = new RevWalk(git.getRepository()).parseCommit(lastCommitId);
+    // Find the node representing the exact file path in the tree.
+    TreeWalk walk = TreeWalk.forPath(git.getRepository(), realRelativePath.toString(), commit.getTree());
+    if (walk == null) {
+      throw new NotFoundException("File not found in Git revision tree.");
+    }
+    return walk.getObjectId(0).getName();
+  }
+
   /**
    * Initializes the Git repository by cloning remote.
    *
+   * @return the commit ID of the present HEAD.
    * @throws GitAPIException               when a Git operation fails.
    * @throws IOException                   when file or network I/O fails.
    * @throws AuthenticationConfigException when there is a failure while fetching authentication credentials for Git.
    */
-  public void cloneRemote() throws Exception {
+  public String cloneRemote() throws Exception {
     if (git != null) {
-      return;
+      return resolveHead();
```

Comment:
This refers to the head of the local repository. This method returns the commit ID of the presently checked out commit. Since no action is performed when `git` is not `null`, the local HEAD should remain unchanged.

Presently we don't require fetching, pulling remote after the initial clone because we clone the repository for every operation. In the future, when we will share the same repository for multiple operations, we can implement pull, fetch.

Reply:
This makes this method work differently depending on if the repo was already cloned or not. If the former case you get the latest HEAD, in the latter case you get some HEAD at some point in time. It's also inconsistent with the method name. And the problematic part is that calling multiple workers would result in HEAD jiggling back and forth in time

Reply:
Imagine a situation: 
 * Operation 1
 * Worker 1: clone repo, a1 branch, head at A
 * Operation 1 finished
 * Commit B is added to git a1 branch
 * Operation 2
 * Worker 2: clone repo, a2 branch, head at B is returned
 * Worker 1: getFileHash(file, B)
 * Worker 1: Can't find commit B, failed
 * Operation 2 failed
Overall, you need to handle this by 
 * Always fetching for clone() to ensure latest remote HEAD is returned
 * Fort other operations either fetch always or handle the use case where commit is not found, then fetch and retry

Reply:
But this can be done in a follow-up, created https://cdap.atlassian.net/browse/CDAP-20386

Reply:
Thanks, I will address this in a follow up PR soon.